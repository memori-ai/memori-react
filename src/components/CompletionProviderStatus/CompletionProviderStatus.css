.memori--completion-provider-status--icon {
    width: 1em;
    height: 1em;
    fill: var(--memori-error-color, red);
}

.memori--completion-provider-status--tooltip.memori-tooltip.memori-tooltip--align-topLeft:not(.memori-tooltip--disabled).memori-tooltip--visible .memori-tooltip--content,
.memori--completion-provider-status--tooltip.memori-tooltip.memori-tooltip--align-topLeft:not(.memori-tooltip--disabled):not(.memori-tooltip--visible):hover .memori-tooltip--content {
    transform: translateY(calc(-100% - 2em)) translateX(3em);
}

.memori--completion-provider-status--tooltip.memori-tooltip .memori-tooltip--content p {
    margin: 0.5em auto;
}

.memori--completion-provider-status--tooltip.memori-tooltip .memori-tooltip--content p+p {
    margin-top: 1em;
}