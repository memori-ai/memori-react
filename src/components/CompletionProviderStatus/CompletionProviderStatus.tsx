import { useEffect, useState } from 'react';
import Tooltip from '../ui/Tooltip';
import Warning from '../icons/Warning';
import { useTranslation } from 'react-i18next';

type Status =
  | 'operational'
  | 'degraded_performance'
  | 'partial_outage'
  | 'major_outage'
  | undefined;

export interface Props {
  forceStatus?: Status;
  provider?: string;
}

const initProviderStatus = (
  provider?: Props['provider'] | 'DEFAULT'
): {
  getStatus: () => Promise<Status>;
  statusPage: string;
} => {
  switch (provider) {
    case 'OpenAI':
      return {
        getStatus: async () => {
          const res = await fetch(
            'https://status.openai.com/api/v2/summary.json'
          );
          const data = await res.json();
          const status = data.components.find(
            (component: { name: string }) => component.name === 'API'
          )?.status as Status;
          return status ?? 'operational';
        },
        statusPage: 'https://status.openai.com/',
      };
    case 'Mistral':
      return {
        getStatus: async () => {
          const res = await fetch(
            'https://status.mistral-data.com/api/v2/summary.json'
          );
          const data = await res.json();
          const status = data.components.find(
            (component: { name: string }) => component.name === 'API'
          )?.status as Status;
          return status ?? 'operational';
        },
        statusPage: 'https://status.mistral-data.com/',
      };
    case 'Anthropic':
      return {
        getStatus: async () => {
          const res = await fetch(
            'https://status.anthropic.com/api/v2/summary.json'
          );
          const data = await res.json();
          const status = data.components.find(
            (component: { name: string }) => component.name === 'API'
          )?.status as Status;
          return status ?? 'operational';
        },
        statusPage: 'https://status.anthropic.com/',
      };
    default:
      return {
        getStatus: async () => 'operational',
        statusPage: '',
      };
  }
};

const CompletionProviderStatus = ({
  forceStatus,
  provider = 'OpenAI',
}: Props) => {
  const { t } = useTranslation();
  const [status, setStatus] = useState<Status>(forceStatus ?? 'operational');

  const providerStatus = initProviderStatus(provider);

  useEffect(() => {
    if (forceStatus) return;

    providerStatus
      .getStatus()
      .then(status => setStatus(status))
      .catch(console.log);
  }, [forceStatus, providerStatus]);

  return status !== 'operational' ? (
    <Tooltip
      className="memori--completion-provider-status--tooltip"
      align="topLeft"
      content={
        <div>
          <p>
            {t('completionProviderDown', {
              provider: provider ?? t('completionProviderFallbackName'),
            })}
          </p>
          {!!providerStatus.statusPage?.length && (
            <p>
              <a
                href={providerStatus.statusPage}
                rel="noopener noreferrer"
                target="_blank"
              >
                {t('completionProviderCheckStatusPage')}
              </a>
            </p>
          )}
        </div>
      }
    >
      <Warning className="memori--completion-provider-status--icon" />
    </Tooltip>
  ) : null;
};

export default CompletionProviderStatus;
