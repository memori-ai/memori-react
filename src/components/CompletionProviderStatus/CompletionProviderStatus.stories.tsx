import React from 'react';
import { Meta, Story } from '@storybook/react';
import I18nWrapper from '../../I18nWrapper';
import CompletionProviderStatus, { Props } from './CompletionProviderStatus';

import './CompletionProviderStatus.css';

const meta: Meta = {
  title: 'Completion Provider Status',
  component: CompletionProviderStatus,
  argTypes: {},
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
};

export default meta;

const Template: Story<Props> = args => (
  <I18nWrapper>
    <CompletionProviderStatus {...args} />
  </I18nWrapper>
);

export const Default = Template.bind({});
Default.args = {};

export const Errored = Template.bind({});
Errored.args = {
  forceStatus: 'major_outage',
};

export const WithSpecifiedProvider = Template.bind({});
WithSpecifiedProvider.args = {
  provider: 'OpenAI',
};

export const ErroredWithSpecifiedProvider = Template.bind({});
ErroredWithSpecifiedProvider.args = {
  forceStatus: 'major_outage',
  provider: 'OpenAI',
};
