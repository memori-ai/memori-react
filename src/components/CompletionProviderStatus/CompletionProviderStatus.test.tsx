import React from 'react';
import { render } from '@testing-library/react';
import CompletionProviderStatus from './CompletionProviderStatus';

it('renders CompletionProviderStatus unchanged', () => {
  const { container } = render(
    <CompletionProviderStatus forceStatus="major_outage" />
  );
  expect(container).toMatchSnapshot();
});

it('renders CompletionProviderStatus errored unchanged', () => {
  const { container } = render(
    <CompletionProviderStatus forceStatus="major_outage" />
  );
  expect(container).toMatchSnapshot();
});

it('renders CompletionProviderStatus with provider specified unchanged', () => {
  const { container } = render(
    <CompletionProviderStatus provider="OpenAI" forceStatus="operational" />
  );
  expect(container).toMatchSnapshot();
});

it('renders CompletionProviderStatus errored with provider specified unchanged', () => {
  const { container } = render(
    <CompletionProviderStatus provider="OpenAI" forceStatus="major_outage" />
  );
  expect(container).toMatchSnapshot();
});
