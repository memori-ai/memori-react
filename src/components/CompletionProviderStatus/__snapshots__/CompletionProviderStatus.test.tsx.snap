// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CompletionProviderStatus errored unchanged 1`] = `
<div>
  <div
    class="memori-tooltip memori-tooltip--align-topLeft memori--completion-provider-status--tooltip"
  >
    <div
      class="memori-tooltip--content"
    >
      <div>
        <p>
          completionProviderDown
        </p>
        <p>
          <a
            href="https://status.openai.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            completionProviderCheckStatusPage
          </a>
        </p>
      </div>
    </div>
    <div
      class="memori-tooltip--trigger"
    >
      <svg
        aria-hidden="true"
        class="memori--completion-provider-status--icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders CompletionProviderStatus errored with provider specified unchanged 1`] = `
<div>
  <div
    class="memori-tooltip memori-tooltip--align-topLeft memori--completion-provider-status--tooltip"
  >
    <div
      class="memori-tooltip--content"
    >
      <div>
        <p>
          completionProviderDown
        </p>
        <p>
          <a
            href="https://status.openai.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            completionProviderCheckStatusPage
          </a>
        </p>
      </div>
    </div>
    <div
      class="memori-tooltip--trigger"
    >
      <svg
        aria-hidden="true"
        class="memori--completion-provider-status--icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders CompletionProviderStatus unchanged 1`] = `
<div>
  <div
    class="memori-tooltip memori-tooltip--align-topLeft memori--completion-provider-status--tooltip"
  >
    <div
      class="memori-tooltip--content"
    >
      <div>
        <p>
          completionProviderDown
        </p>
        <p>
          <a
            href="https://status.openai.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            completionProviderCheckStatusPage
          </a>
        </p>
      </div>
    </div>
    <div
      class="memori-tooltip--trigger"
    >
      <svg
        aria-hidden="true"
        class="memori--completion-provider-status--icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders CompletionProviderStatus with provider specified unchanged 1`] = `<div />`;
